src/VisualStudio/Core/Def/Implementation/Diagnostics/SolutionCrawlerEventHandlerWithDelay.cs

```diff
@@ -0,0 +1,124 @@
+            /// Task used to invoke events outside the <see cref="_lock"/>
+            /// </summary>
+            private Task _invokeEventTask;
+
+            /// <summary>
+            /// Task used to trigger event invocation on an interval
+            /// defined by <see cref="_publishInterval"/>
+            /// </summary>
+            private Task _intervalTask;
+
+            private bool _isTimerRunning;
+
+            /// <summary>
+            /// Solution crawler events throttled to the specified <see cref="_publishInterval"/>
+            /// </summary>
+            public EventHandler<ProgressData>? UpdateProgress;
+
+            public SolutionCrawlerEventHandlerWithDelay(ISolutionCrawlerProgressReporter reporter, TimeSpan publishInterval)
+            {
+                _publishInterval = publishInterval;
+                _invokeEventTask = Task.CompletedTask;
+                _intervalTask = Task.CompletedTask;
+                _isTimerRunning = false;
+
+                // no event unsubscription since it will remain alive until VS shutdown
+                reporter.ProgressChanged += OnSolutionCrawlerProgressChanged;
+            }
+
+            /// <summary>
+            /// Retrieve and throttle solution crawler events to be sent to the progress reporter UI.
+            /// 
+            /// there is no concurrent call to this method since ISolutionCrawlerProgressReporter will serialize all
+            /// events to preserve event ordering
+            /// </summary>
+            /// <param name="progressData"></param>
+            public void OnSolutionCrawlerProgressChanged(object sender, ProgressData progressData)
+            {
+                lock (_lock)
+                {
+                    _progressData = progressData;
+
+                    // The publish event timer is already running. If this is a stop event make sure the timer does not get extended.
+                    // Otherwise, just update the progress data and continue.
+                    if (_isTimerRunning)
+                    {
+                        if (progressData.Status == ProgressStatus.Stopped)
+                        {
+                            _isTimerRunning = false;
+                        }
+                        else
+                        {
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        // Event publishing is not running, so trigger new timer.
+                        _isTimerRunning = true;
+                        RunOnTimer();
+                    }
+                }
+            }
+
+            private void InvokeEvent()
+            {
+                _invokeEventTask = _invokeEventTask.SafeContinueWith(_ =>
+                {
+                    // No need to publish the exact same event over and over.
+                    if (!Equals(_progressData, _lastPublishedProgressData))
```

Comment:
in my model this is much simpler  when `SomeEvent` calls in with the _progressData you just enter the lock, see if the value you have is the same as the value from before, if so, you don't need to do anything as the previous caller will have ensured all teh right work is kicked off.